mapie/mondrian.py

```diff
@@ -0,0 +1,438 @@
+        """
+        Fit the Mondrian method
+
+        Parameters
+        ----------
+        X : ArrayLike of shape (n_samples, n_features)
+            The input data
+        y : ArrayLike of shape (n_samples,) or (n_samples, n_outputs)
+            The target values
+        groups : ArrayLike of shape (n_samples,)
+            The groups of individuals
+        **kwargs
+            Additional keyword arguments to pass to the estimator's fit method
+            that may be specific to the Mapie estimator used
+        """
+
+        X, y, groups = self._check_fit_parameters(X, y, groups)
+        self.unique_groups = np.unique(groups)
+        self.mapie_estimators = {}
+
+        for group in self.unique_groups:
+            mapie_group_estimator = deepcopy(self.mapie_estimator)
+            indices_groups = np.argwhere(groups == group)[:, 0]
+            X_g, y_g = X[indices_groups], y[indices_groups]
+            mapie_group_estimator.fit(X_g, y_g, **kwargs)
+            self.mapie_estimators[group] = mapie_group_estimator
+        return self
+
+    def predict(
+            self, X: ArrayLike, groups: ArrayLike,
+            alpha: Optional[Union[float, Iterable[float]]] = None,
+            **kwargs
+    ) -> Union[NDArray, Tuple[NDArray, NDArray]]:
+        """
+        Perform conformal prediction for each group of individuals
+
+        Parameters
+        ----------
+        X : ArrayLike of shape (n_samples, n_features)
+            The input data
+        groups : ArrayLike of shape (n_samples,)
+            The groups of individuals
+        alpha : float or Iterable[float], optional
+            The desired coverage level(s) for each group.
+
+            By default None.
+        **kwargs
+            Additional keyword arguments to pass to the estimator's predict
+            method that may be specific to the Mapie estimator used
+
+        Returns
+        -------
+        y_pred : NDArray of shape (n_samples,) or (n_samples, n_outputs)
+            The predicted values
+        y_pss : NDArray of shape (n_samples, n_outputs, n_alpha)
+        """
+
+        check_is_fitted(self, self.fit_attributes)
+        self._check_not_topk_calibrator()
+        X = cast(NDArray, X)
+        groups = self._check_groups_predict(X, groups)
+        if alpha is None and self.mapie_estimator.estimator is not None:
+            return self.mapie_estimator.estimator.predict(X, **kwargs)
+        else:
+            alpha_np = cast(NDArray, check_alpha(alpha))
+            unique_groups = np.unique(groups)
+            for i, group in enumerate(unique_groups):
+                m = self.mapie_estimators[group]
+                indices_groups = np.argwhere(groups == group)[:, 0]
+                X_g = X[indices_groups]
+                pred = m.predict(X_g, alpha=alpha_np, **kwargs)  # type: ignore
+                y_pred_g, y_pss_g = pred
+                if i == 0:
```

Comment:
Using the first loop iteration to check for alpha dimensionality is convoluted. Doing this before the loop, maybe using existing utils if needed, seems like a better option. I can spend some time finding a proper solution.

Reply:
done